### PR TITLE
[Balloon] Add integration test for memory scrubbing

### PR DESCRIPTION
**Pull request by georgepisaltu**
## Reason for This PR

We needed an integration test to ensure memory allocated after an inflate/deflate cycle is indeed zeroed out.

## Description of Changes

Added a test for the ballon device that ensures the memory after a deflate is zeroed out.

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] The reason for this PR is clearly provided (issue no. or explanation).
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any API changes are reflected in `firecracker/swagger.yaml`.
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.

### CODE REVIEW

File: tests/host_tools/readmem.c [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            PROT_READ | PROT_WRITE,
+            MAP_ANONYMOUS | MAP_PRIVATE,
+            -1,
+            0
+        );
+    } while (ptr == MAP_FAILED);
+        
+    cur = (int *) ptr;
+    for(i = 0; i < (mb_count * MB * sizeof(char)) / (4 * sizeof(int)); i += 4) {
+        if (cur[i] == value &&
```
Comment (gc-plp) [inline tests/host_tools/readmem.c:32]: `value` could be used to fill a buffer and do the comparison using `memcmp`.

File: tests/host_tools/readmem.c [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            mb_count * MB * sizeof(char),
+            PROT_READ | PROT_WRITE,
+            MAP_ANONYMOUS | MAP_PRIVATE,
+            -1,
+            0
+        );
+    } while (ptr == MAP_FAILED);
+        
+    cur = (int *) ptr;
+    for(i = 0; i < (mb_count * MB * sizeof(char)) / (4 * sizeof(int)); i += 4) {
```
Comment (gc-plp) [inline tests/host_tools/readmem.c:31]: We should explain the logic here with a comment.

File: tests/host_tools/readmem.c (hunk 1 above)
Comment (gc-plp) [inline tests/host_tools/readmem.c:32]: If I understand this correctly, this loop returns `1` when a memory area consisting of 4 consecutive `int`s are equal to `value`.

You could make it quicker by filling a buffer with `value` and using `memcmp` to compare where `cur` points to the buffer.

File: tests/host_tools/readmem.c [hunk 3]
```diff
@@ -0,0 +1,55 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
```
Comment (gc-plp) [inline tests/host_tools/readmem.c:3]: What is the overall purpose of this application?

Comment (georgepisaltu) [reply tests/host_tools/readmem.c:3]: Added a description of the program.

File: tests/host_tools/readmem.c (hunk 2 above)
Comment (georgepisaltu) [reply tests/host_tools/readmem.c:31]: Added the comment.

File: tests/host_tools/readmem.c (hunk 1 above)
Comment (georgepisaltu) [reply tests/host_tools/readmem.c:32]: That's true, I changed it to use a buffer and `memcmp`.

**Review by gc-plp — DISMISSED** [review]

File: tests/integration_tests/functional/test_balloon.py [hunk 4]
```diff
@@ -700,3 +706,90 @@ def _test_snapshot_compatibility(context):
     )
 
     microvm.kill()
+
+
+@pytest.mark.skipif(
```
Comment (dianpopa) [inline tests/integration_tests/functional/test_balloon.py:711]: Why?

File: tests/integration_tests/functional/test_balloon.py [hunk 5]
```diff
@@ -700,3 +706,90 @@ def _test_snapshot_compatibility(context):
...
+    platform.machine() != "x86_64",
+    reason="Not supported yet."
+)
+def test_memory_scrub(
+    network_config,
+    bin_cloner_path
+):
+    """Test that the memory is zeroed after deflate."""
+    logger = logging.getLogger("snapshot_sequence")
```
Comment (dianpopa) [inline tests/integration_tests/functional/test_balloon.py:720]: Not really understanding the need to use snapshotting for making sure memory is zeroed out after deflate?

File: tests/integration_tests/functional/test_balloon.py (hunk 4 above)
Comment (georgepisaltu) [reply tests/integration_tests/functional/test_balloon.py:711]: Forgot to remove the annotation. It's gone now.

File: tests/integration_tests/functional/test_balloon.py (hunk 5 above)
Comment (georgepisaltu) [reply tests/integration_tests/functional/test_balloon.py:720]: It was the wrong logger call. Now it uses `logging.getLogger()`, which returns the root logger.

**Review by dianpopa — APPROVED** [review]

**Review by gc-plp — APPROVED** [review]

**Pull request merged**